File: src/java/magma/app/compile/rule/SymbolRule.java
package magma.app.compile.rule;

import magma.api.collect.head.HeadedIter;
import magma.api.collect.head.RangeHead;
import magma.api.option.None;
import magma.api.option.Option;
import magma.api.text.Characters;
import magma.api.text.Strings;
import magma.app.compile.node.Node;

public record SymbolRule(TypeRule childRule) implements Rule<Node> {
    public static boolean isSymbol(String input) {
        var query = new HeadedIter<Integer>(new RangeHead(Strings.length(input)));
        return query.allMatch((Integer index) -> {
            return SymbolRule.isSymbolChar(index, input.charAt(index));
        });
    }

    private static boolean isSymbolChar(int index, char c) {
        return '_' == c
                || Characters.isLetter(c)
                || (0 != index && Characters.isDigit(c));
    }

    @Override
    public Option<Node> lex(String input) {
        if (SymbolRule.isSymbol(input)) {
            return this.childRule().lex(input);
        }

        return new None<>();
    }
}

File: src/ts/magma/app/compile/rule/SymbolRule.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Rule } from "../../../../magma/app/compile/rule/Rule";
import { TypeRule } from "../../../../magma/app/compile/rule/TypeRule";
import { HeadedIter } from "../../../../magma/api/collect/head/HeadedIter";
import { RangeHead } from "../../../../magma/api/collect/head/RangeHead";
import { Strings } from "../../../../magma/api/text/Strings";
import { Characters } from "../../../../magma/api/text/Characters";
import { Option } from "../../../../magma/api/option/Option";
import { None } from "../../../../magma/api/option/None";
export class SymbolRule implements Rule<Node> {
	childRule: TypeRule;
	constructor (childRule: TypeRule) {
		this.childRule = childRule;
	}
	static isSymbol(input: string): boolean {
		let query = new HeadedIter<number>(new RangeHead(Strings.length(input)))/*unknown*/;
		return query.allMatch((index: number) => {
			return SymbolRule.isSymbolChar(index, input.charAt(index))/*unknown*/;
		})/*unknown*/;
	}
	static isSymbolChar(index: number, c: string): boolean {
		return "_" === c || Characters.isLetter(c) || (0 !== index && Characters.isDigit(c))/*unknown*/;
	}
	lex(input: string): Option<Node> {
		if (SymbolRule.isSymbol(input)/*unknown*/){
			return this.childRule().lex(input)/*unknown*/;
		}
		return new None<?>()/*unknown*/;
	}
}
